eval/src/vespa/eval/tensor/partial_update.cpp

```diff
@@ -0,0 +1,388 @@
+                offset += coord;
+            }
+        } else {
+            // "bad label{%s} in modifier tensor, was %zu, must be < %zu", dim_names[dim], coord, cur
+            offset = npos();
+        }
+        ++dim;
+    }
+    void with(vespalib::stringref label) {
+        uint32_t result = 0;
+        for (char c : label) {
+            if (c < '0' || c > '9') { // bad char
+                // "bad label{%s} in modifier tensor, was '%s'", dim_names[dim], label.data()
+                offset = npos();
+                ++dim;
+                return;
+            }
+            result = result * 10 + (c - '0');
+        }
+        with(result);
+    }
+    void add_dim(const char *name, size_t sz) {
+        dim_sizes.push_back(sz);
+        dim_names.push_back(name);
+        total_size *= sz;
+    }
+    size_t get() const {
+        assert(dim == dim_sizes.size());
+        return offset;
+    }
+    ~DenseCoords();
+};
+DenseCoords::~DenseCoords() = default;
+
+struct Addresses {
+    std::vector<vespalib::stringref> addr;
+    std::vector<vespalib::stringref *> next_result_refs;
+    std::vector<const vespalib::stringref *> lookup_refs;
+    std::vector<size_t> lookup_view_dims;
+    Addresses(size_t sz)
+        : addr(sz), next_result_refs(sz), lookup_refs(sz), lookup_view_dims(sz)
+    {
+        for (size_t i = 0; i < sz; ++i) {
+            next_result_refs[i] = &addr[i];
+            lookup_refs[i] = &addr[i];
+            lookup_view_dims[i] = i;
+        }
+    }
+    ~Addresses();
+};
+Addresses::~Addresses() = default;
+
+struct AddressHandler {
+    std::vector<DimCase> how;
+    DenseCoords target_coords;
+    Addresses for_output;
+    Addresses from_modifier;
+    bool valid;
+
+    AddressHandler(const ValueType &input_type,
+                const ValueType &modifier_type)
+        : how(), target_coords(),
+          for_output(input_type.count_mapped_dimensions()),
+          from_modifier(modifier_type.count_mapped_dimensions()),
+          valid(true)
+    {
+        if (! modifier_type.is_sparse()) {
+            LOG(error, "Unexpected non-sparse modifier tensor, type is %s",
+                modifier_type.to_spec().c_str());
+            valid = false;
+            return;
+        }
+        // analyse dimensions
+        auto visitor = overload {
+            [&](visit_ranges_either, const auto &) { valid = false; },
+            [&](visit_ranges_both, const auto &a, const auto &) {
+                how.push_back(a.is_mapped() ? DimCase::MAPPED_MATCH : DimCase::CONV_TO_INDEXED);
+            }
+        };
+        const auto & input_dims = input_type.dimensions();
+        const auto & modifier_dims = modifier_type.dimensions();
+        visit_ranges(visitor,
+                     input_dims.begin(), input_dims.end(),
+                     modifier_dims.begin(), modifier_dims.end(),
+                     [](const auto &a, const auto &b){ return (a.name < b.name); });
+        if ((! valid) ||
```

Comment:
consider checking number of dimensions together with modifier sparseness and only checking valid here.